src/shogun/solver/LDASolver.h

```diff
@@ -110,48 +111,44 @@ namespace shogun
 	void LDASolver<T>::compute_means()
 	{
 		index_t num_class = m_labels->get_num_classes();
-		auto data = m_features->get_feature_matrix();
 
 		m_class_mean = std::vector<SGVector<T>>(num_class);
 		m_class_count = std::vector<index_t>(num_class);
 		for (index_t i = 0; i < num_class; ++i)
 		{
-			m_class_mean[i] = SGVector<T>(data.num_rows);
+			m_class_mean[i] = SGVector<T>(m_features->get_num_features());
 			linalg::zero(m_class_mean[i]);
 		}
-		m_mean = SGVector<T>(data.num_rows);
+		m_mean = SGVector<T>(m_features->get_num_features());
 		linalg::zero(m_mean);
 
-		// calculate the total mean and the classes' mean.
-		for (index_t i = 0; i < data.num_cols; ++i)
+		for (const auto& [data_i, label_i]: zip_iterator(m_features, m_labels))
 		{
-			index_t c = (index_t)m_labels->get_label(i);
+			auto c = static_cast<index_t>(label_i);
 			++m_class_count[c];
-			linalg::add_col_vec(data, i, m_class_mean[c], m_class_mean[c]);
+			linalg::add(data_i, m_class_mean[c], m_class_mean[c]);
 		}
+
 		for (index_t i = 0; i < num_class; ++i)
 		{
 			linalg::add(m_mean, m_class_mean[i], m_mean);
 			linalg::scale(
 			    m_class_mean[i], m_class_mean[i], 1 / (T)m_class_count[i]);
 		}
-		linalg::scale(m_mean, m_mean, 1 / (T)data.num_cols);
+		linalg::scale(m_mean, m_mean, 1 / (T)m_features->get_num_vectors());
 	}
 
 	template <typename T>
 	void LDASolver<T>::compute_within_cov()
 	{
 		index_t num_features = m_features->get_num_features();
-		index_t num_vectors = m_features->get_num_vectors();
 		index_t num_class = m_labels->get_num_classes();
 
-		auto data = m_features->get_feature_matrix().clone();
+		auto data = static_cast<CDenseFeatures<T>*>(m_features->clone());
 
 		// Center data with respect to each data point's class
-		for (index_t i = 0; i < data.num_cols; ++i)
-			linalg::add_col_vec(
-			    data, i, m_class_mean[m_labels->get_label(i)], data, (T)1.0,
-			    (T)-1.0);
+		for (auto [data_i, label_i]: zip_iterator(data, m_labels))
+			linalg::add(data_i, m_class_mean[label_i], data_i, static_cast<T>(1), static_cast<T>(-1));
```

Comment:
@karlnapf I am not sure if this is any nicer than before? I guess it's a matter of taste! In any case, speed wise it'll be the same

Reply:
I like it

Reply:
Does this write into the feature vector? Because if so, no good! Shouldn't be allowed
If that is the case, we need to clone/copy the matrix above explicitly (instead of cloning the features)

Reply:
To not be able to write to the original data you need to do `const auto&`. Ah i see, but doesn't the clone of features clone the matrix?

Reply:
we want to avoid writing to the output of the iteratior definitely imo. that is by design of the iterator...otherwise people will use it.

and i am not sure why clone the features? not necessary